### ISSUE TITLE
General IDE lagginess 

### ISSUE THREAD

**Issue by cherryleafroad** [intellij-rust/intellij-rust#8966]
## Environment

* **IntelliJ Rust plugin version:** 0.4.172.4687-221
* **Rust toolchain version:** 1.61.0 (fe5b13d68 2022-05-18) x86_64-pc-windows-msvc
* **IDE name and version:** IntelliJ IDEA 2022.1 Ultimate Edition (IU-221.5080.210)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.macros.proc

## Problem description
Recently, code completion, and even general IDE experience often greatly lags. Code completion for something as simple as a variable name in the same scope can sometimes take 20 seconds to pop up.

Includes generally any other experience, like clicking to go to the source file of a reference (although the lagginess does depend on the feature at hand).

Cargo Check being turned on or off doesn't change much (though of course cargo check certainly is laggier for obvious reasons; unsure if it's a contributory factor as I often turn it on and off as needed).

Of course, this testing was done after indexing and everything was already finished.

_The issue doesn't always happen, and changes with usage_, so sometimes you may see code completions being snappy enough, but when you really get into your programming sweetspot where you're really doing stuff, browsing the docs that pop up on hover, adding libs to cargo (and letting it finish indexing), etc, it pops up fairly frequently during real sessions (which gets you out of the zone).

- It is unknown if it's the IDE's fault itself, but I seem to remember in the past using this plugin wasn't so slow.

- I have made every effort to make sure the IDE itself is getting what it needs to be fast. E.g. increasing VM memory, disabling unused plugins, turning off `import on the fly`-features. My `idea64.exe.vmoptions` config is below, just in case that's impacting something negatively.

`idea64.exe.vmoptions`:
<details>
  <summary>Click to expand!</summary>

```
-Xms1024m
-Xmx8192m
-Xss64m
-XX:NewRatio=2
-Dfile.encoding=UTF-8
-XX:SoftRefLRUPolicyMSPerMB=250
-XX:NewSize=512m
-XX:MaxNewSize=512m
-XX:PermSize=512m
-XX:MaxPermSize=1024m
-XX:+UseParNewGC
-XX:ParallelGCThreads=4
-XX:MaxTenuringThreshold=1
-XX:SurvivorRatio=8
-XX:+UseCodeCacheFlushing
-XX:+AggressiveOpts
-XX:+CMSClassUnloadingEnabled
-XX:+CMSIncrementalMode
-XX:+CMSIncrementalPacing
-XX:+CMSParallelRemarkEnabled
-XX:CMSInitiatingOccupancyFraction=65
-XX:+CMSScavengeBeforeRemark
-XX:+UseCMSInitiatingOccupancyOnly
-XX:-TraceClassUnloading
-XX:+AlwaysPreTouch
-XX:+TieredCompilation
-XX:+DoEscapeAnalysis
-XX:+UnlockExperimentalVMOptions
-XX:LargePageSizeInBytes=256m
-XX:+DisableExplicitGC
-XX:+ExplicitGCInvokesConcurrent
-XX:+PrintGCDetails
-XX:+PrintFlagsFinal
-XX:+CMSPermGenSweepingEnabled
-XX:+UseAdaptiveGCBoundary
-XX:+UseSplitVerifier
-XX:CompileThreshold=10000
-XX:+UseCompressedStrings
-XX:+OptimizeStringConcat
-XX:+UseStringCache
-XX:+UseFastAccessorMethods
-XX:+UnlockDiagnosticVMOptions
-Djava.net.preferIPv4Stack=true

-XX:ErrorFile=$USER_HOME/java_error_in_idea_%p.log
-XX:HeapDumpPath=$USER_HOME/java_error_in_idea.hprof
-Dide.no.platform.update=true
```

</details>

## Steps to reproduce

- None, as source is unknown, however, I've provided as much detail as I can in the description section.

Appears to be similar or the same as this issue here: https://github.com/intellij-rust/intellij-rust/issues/8920

**Comment by dima74** [comment]
@cherryleafroad could you please try previous version [0.4.170](https://plugins.jetbrains.com/plugin/8182-rust/versions/stable/175114) and check whether performance issues persist? This will help us investigate the issue

**Comment by TobiasDeBruijn** [comment]
> My project is large (about 4k lines of code), so a large codebase may be a contributory factor as well.
Yep this definitely. Working on some smaller projects as well, the IDE is much, much faster. I only have issues on larger projects (~12K lines)

**Comment by iangregsondev** [comment]
I just tried to install 0.4.170 has suggested above..

i use a lot of Jetbrains products, and normally use each product for there specific job. So hence I use Clion for RUST..

CPU through the roof, the machine is almost dead .. I have a Mac M1 Pro ...

I though it might be CLION so I tired IDEA and same as soon as I install the plugin and open a small RUST app, the IDE just freezes and CPU is soooooo high

![image](https://user-images.githubusercontent.com/2370364/175602363-bd8a52b9-d23a-4c47-8617-cd7c080532fc.png)

![image](https://user-images.githubusercontent.com/2370364/175602476-16b248cd-2374-404b-a01d-8c447a8e1c81.png)

**Comment by iangregsondev** [comment]
Its the same in either CLION or IDEA.. I think its related to the plugin..

**Comment by dima74** [comment]
> I though it might be CLION so I tired IDEA and same as soon as I install the plugin and open a small RUST app, the IDE just freezes and CPU is soooooo high

@iangregsondev When you install update of the plugin (or downgrade like in this case), and first-time launch the IDE, the plugin does a lot of job like indexing, etc. This is done one time, and any subsequent launches should be much faster. Indexing can take 1-2 minutes, and it is expected. After indexing code highlighting, completion, etc expected to be fast. So we ask you to compare e.g. completion performance of v170 and latest version

**Comment by iangregsondev** [comment]
Yes from what I can see, first time I start Clion/IDEA then the CPU is way way up there... but if I just wait.... it drops down.. but I open a .RS file and it shoots up again.

Although its not consistent, it shoots up to 700%, then I kept testing it and the next time only 89%

I will try and compare them later.

Thank you.

**Comment by iangregsondev** [comment]
Here is my plugin version fyi. I haven't updated it as I am using the previous version as per above..

![image](https://user-images.githubusercontent.com/2370364/175609751-1da73d9d-3f95-45ab-9491-f77909f64f55.png)

**Comment by PepejHa4ker** [comment]
The same problem.

**Comment by HolgerGottChristensen** [comment]
I am experiencing similar issues with later versions of the plugin.

I tried different versions of the plugin:
0.4.176
0.4.172.4687-222
0.4.170.4627-221
0.4.169.4584-221
0.4.168.4555-221

The lag stops when downgrading to: 0.4.167.4525-221

Here is my environment.

## Environment
* **IntelliJ Rust plugin version:** 0.4.167.4525-221
* **Rust toolchain version:** 1.62.0 (a8314ef7d 2022-06-27) aarch64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2022.2.1 Ultimate Edition (IU-222.3739.54)
* **Operating system:** macOS 12.4
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc


I noticed from 0.4.168 it started saying something like "saving macro expansions" in the progress thing at the bottom of the IDE, but it never said expanding macros, like it did in the earlier versions. In the earlier versions it doesn't say "saving macro expansions" (or maybe it is just so fast that I didn't notice).

The performance also breaks syntax highlighting for me.

**Comment by vlad20012** [comment]
@HolgerGottChristensen Hi! Thank you for the investigation! 
Could you please also try the latest nightly plugin version? The issue can be fixed by #9208 that is currently in the nightly only. See [instructions](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly) on how to install the the plugin from the nightly channel

**Comment by HolgerGottChristensen** [comment]
Running on: 0.4.178.4836-222-nightly did not fix the issues for me.

**Comment by vlad20012** [comment]
If you work on some open-source project, could you please link it and say in which files do you experience lagginess?

**Comment by HolgerGottChristensen** [comment]
I am working in this repository
https://github.com/HolgerGottChristensen/carbide

The lagginess is quite general, but an example is this file:
carbide_core/src/widget/shape/capsule.rs

An example of a file with syntax highlighting on 0.4.167.4525-221
<img width="919" alt="Screenshot 2022-08-29 at 12 18 04" src="https://user-images.githubusercontent.com/11473146/187179666-d8813089-f014-4ced-8e3b-5cfe88e3d105.png">

and when updating to 0.4.178.4836-222-nightly:
<img width="826" alt="Screenshot 2022-08-29 at 12 20 32" src="https://user-images.githubusercontent.com/11473146/187180014-d6ffac60-ca37-453e-b4fb-af124e9198d4.png">

and some syntax highlighting starts to show up after 30sec, but sporadically
<img width="723" alt="Screenshot 2022-08-29 at 12 21 52" src="https://user-images.githubusercontent.com/11473146/187180278-46939ede-c545-4403-bca3-120a024b51d1.png">

**Comment by TobiasDeBruijn** [comment]
On the current nightly version `0.4.177.4809-221-nightly` the completion performance seems to be a bit worse.
CPU is not maxed out at all, though. Note that the performance only becomes a problem on large projects.

**Comment by vlad20012** [comment]
@HolgerGottChristensen, Your issue should be fixed in the nightly plugin. You can try it now (see [instructions](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly)).

That actually was not a performance issue. Highlighting just stop at some point because of the bug with macro expansion. I think I (hot)fixed the bug in #9280, and the fix is now released in the nightly plugin channel.

**Comment by HolgerGottChristensen** [comment]
> @HolgerGottChristensen, Your issue should be fixed in the nightly plugin. You can try it now (see [instructions](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly)).
> 
> That actually was not a performance issue. Highlighting just stop at some point because of the bug with a macro expansion (the bug was introduced in `0.4.168` release because we rewritten the macro expansion code). I think I (hot)fixed the bug in #9280, and the fix is now released in the nightly plugin channel.

Great, thank you! 

I am seeing some false positive error messages, but those are definitely not performance related, and should probably be a separate issue.

**Comment by vlad20012** [comment]
@MolotovCherry, do you still experience "General IDE lagginess" as you stated in the original issue on the latest plugin version?

**Comment by MolotovCherry** [comment]
> @MolotovCherry, do you still experience "General IDE lagginess" as you stated in the original issue on the latest plugin version?

It seems to be better than before!

**Comment by vlad20012** [comment]
@MolotovCherry, Glad to hear! We're working hard on performance issues last months, but still we have a lot to do in this direction. If you work on some open-source project, could you please link it here along with a file in that project where you particularly experiencing lagginess? Then we could better test our plugin performance against your project.

**Comment by MolotovCherry** [comment]
> @MolotovCherry, Glad to hear! We're working hard on performance issues last months, but still we have a lot to do in this direction. If you work on some open-source project, could you please link it here along with a file in that project where you particularly experiencing lagginess? Then we could better test our plugin performance against your project.

Sure thing! One of my projects I most worked on is [this one](https://github.com/MolotovCherry/kmagick), where [this folder](https://github.com/MolotovCherry/kmagick/tree/main/rust/jni-macros/src) contains a ton of proc macro code, and [this file](https://github.com/MolotovCherry/kmagick/blob/main/rust/kmagick/src/magick_wand.rs) contains heavy use of macros.

**Comment by MolotovCherry** [comment]
Closing as this project is deprecated
